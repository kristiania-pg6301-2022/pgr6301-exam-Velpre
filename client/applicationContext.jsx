import React from "react";
import { fetchJSON, postJSON } from "./lib/http";

export const ApplicationContext = React.createContext({
  async fetchLogin() {
    return await fetchJSON("/api/login");
  },

  async registerLogin(provider, login) {
    return await postJSON(`/api/login/${provider}`, login);
  },
  async listArticles() {
    return fetchJSON(`/api/articles`);
  },

  async createArticle(article) {
    return await postJSON("/api/articles/new", article);
  },

  async deleteArticle(article) {
    return await postJSON("/api/articles/delete", article);
  },

  async updateArticle(article) {
    return await postJSON("/api/articles/update", article);
  },

  async logout(url, reload, navigate) {
    await fetch(url, { method: "delete" });
    reload();
    navigate("/");
  },
});
