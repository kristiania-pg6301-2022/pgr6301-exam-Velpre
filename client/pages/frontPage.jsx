import { Link } from "react-router-dom";
import React from "react";

export function FrontPage() {
    //skal liste bare titell fra DB med en gang
  return (
    <div>
      <h1>VÃ¥re Artikler</h1>
      <ul>
        <li>
          <Link to={"/movies"}>List movies</Link>
        </li>
      </ul>
    </div>
  );
}
