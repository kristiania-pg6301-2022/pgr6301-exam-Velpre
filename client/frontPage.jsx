import React, { useState, useContext } from "react";
import { useLoading } from "./lib/useLoading";
import { ApplicationContext } from "./applicationContext";

export function FrontPage({ user, reload }) {
  return (
    <div>
      <ListArticles user={user} />
      {user.hk && (
        <div className="front-page-editor">
          <EditorAdd />
          <EditorUpdate />
        </div>
      )}
    </div>
  );
}

export function ListArticles({ user }) {
  const { listArticles } = useContext(ApplicationContext);
  const [chosenArticle, setChosenArticle] = useState("");

  const { loading, error, data } = useLoading(async () => listArticles(), []);

  if (loading) {
    return <div>Loading...</div>;
  }
  if (error) {
    return (
      <div>
        <h1>Error</h1>
        <div id="error-text">{error.toString()}</div>
      </div>
    );
  }

  const handleClick = (article) => {
    setChosenArticle(article);
  };

  if (user.google || user.hk) {
    return (
      <div className="list-articles-frontpage-loged">
        <h1>Our Articles</h1>
        <div className="grid-split-frontpage">
          <div>
            {data.map((article, index) => (
              <div key={index}>
                <button onClick={() => handleClick(article)}>
                  {article.title}
                </button>
              </div>
            ))}
          </div>
          <div>{chosenArticle && <ArticleCard article={chosenArticle} />}</div>
        </div>
      </div>
    );
  }

  return (
    <div className="list-articles-frontpage">
      <h1>Our Articles</h1>
      {data.map((article, index) => (
        <div>
          <h3>{article.title}</h3>
        </div>
      ))}
    </div>
  );
}

function ArticleCard({ article }) {
  return (
    <div>
      <h1>{article.title}</h1>
      <p>Category: {article.category}</p>
      <p>Plot: {article.plot}</p>
    </div>
  );
}

export function EditorAdd() {
  const { createArticle } = useContext(ApplicationContext);
  const [title, setTitle] = useState("");
  const [category, setCategory] = useState("");
  const [plot, setPlot] = useState("");

  async function handleSubmit(e) {
    e.preventDefault();
    createArticle({ title, category, plot });
    setTitle("");
    setCategory("");
    setPlot("");
  }

  return (
    <form onSubmit={handleSubmit}>
      <h1>Add Article</h1>
      <div>
        <label>
          <div>
            <strong>Category:</strong>
          </div>
          <select required onChange={(e) => setCategory(e.target.value)}>
            <option value={"General"}>General</option>
            <option value={"Local"}>Local</option>
            <option value={"Aboard"}>Aboard</option>
          </select>
        </label>
      </div>
      <div>
        <label>
          <div>
            <strong>Title:</strong>
          </div>
          <input
            required
            value={title}
            onChange={(e) => setTitle(e.target.value)}
          />
        </label>
      </div>
      <div>
        <label>
          <div>
            <strong>Plot:</strong>
          </div>
          <textarea
            required
            value={plot}
            onChange={(e) => setPlot(e.target.value)}
          />
        </label>
      </div>

      <button>Submit</button>
    </form>
  );
}

function EditorUpdate() {
  const { createArticle } = useContext(ApplicationContext);
  const [title, setTitle] = useState("");
  const [category, setCategory] = useState("");
  const [plot, setPlot] = useState("");

  //const navigate = useNavigate();

  async function handleSubmit(e) {
    e.preventDefault();
    const res = await createArticle({ title, category, plot });
    console.log(res);

    setTitle("");
    setCategory("");
    setPlot("");
  }

  return (
    <form onSubmit={handleSubmit}>
      <h1>Add Article</h1>
      <div>
        <label>
          <div>
            <strong>Category:</strong>
          </div>
          <select required onChange={(e) => setCategory(e.target.value)}>
            <option value={"General"}>General</option>
            <option value={"Local"}>Local</option>
            <option value={"Aboard"}>Aboard</option>
          </select>
        </label>
      </div>
      <div>
        <label>
          <div>
            <strong>Title:</strong>
          </div>
          <input
            required
            value={title}
            onChange={(e) => setTitle(e.target.value)}
          />
        </label>
      </div>
      <div>
        <label>
          <div>
            <strong>Plot:</strong>
          </div>
          <textarea
            required
            value={plot}
            onChange={(e) => setPlot(e.target.value)}
          />
        </label>
      </div>

      <button>Submit</button>
    </form>
  );
}
