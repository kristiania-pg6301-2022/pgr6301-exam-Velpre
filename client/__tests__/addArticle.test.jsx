import React from "react";
import ReactDOM from "react-dom";
import { Simulate } from "react-dom/test-utils";
import { ApplicationContext } from "../applicationContext";
import { EditorAdd } from "../frontPage";

describe("add article", () => {
  it("shows article form", () => {
    const element = document.createElement("div");
    ReactDOM.render(<EditorAdd />, element);
    expect(element.innerHTML).toMatchSnapshot();
    expect(
      Array.from(element.querySelectorAll("form label strong")).map(
        (e) => e.innerHTML
      )
    ).toEqual(["Category:", "Title:", "Plot:"]);
    expect(element.querySelector("h1").innerHTML).toEqual("Add Article");
  });

  it("adds Articles on submit", () => {
    const createArticle = jest.fn();
    const handleNewArticle = jest.fn();
    const element = document.createElement("div");
    ReactDOM.render(
      <ApplicationContext.Provider value={{ createArticle, handleNewArticle}}>
        <EditorAdd handleNewArticle={handleNewArticle}/>
      </ApplicationContext.Provider>,
      element
    );

    Simulate.change(element.querySelector("form div:nth-of-type(1) select"), {
      target: { value: "General" },
    });
    Simulate.change(element.querySelector("form div:nth-of-type(2) input"), {
      target: { value: "Article title" },
    });

    Simulate.submit(element.querySelector("form"));
    expect(createArticle).toBeCalledWith({
      category: "General",
      title: "Article title",
      plot: "",
    });
  });
});
