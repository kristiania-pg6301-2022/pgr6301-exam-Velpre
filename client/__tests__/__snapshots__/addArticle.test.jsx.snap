// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add article shows article form 1`] = `"<form><h1>Add Article</h1><div><label><div><strong>Category:</strong></div><select required=\\"\\"><option value=\\"General\\">General</option><option value=\\"Local\\">Local</option><option value=\\"Aboard\\">Aboard</option></select></label></div><div><label><div><strong>Title:</strong></div><input required=\\"\\" value=\\"\\"></label></div><div><label><div><strong>Plot:</strong></div><textarea required=\\"\\"></textarea></label></div><button>Submit</button></form>"`;
